File: sol/solution/0001-0100/0030. Substring with Concatenation of All Words/Solution.go
func findSubstring(s string, words []string) []int {
    if len(words) == 0 {
        return []int{}
    }
    wordLen := len(words[0])
    wordCount := make(map[string]int)
    for _, w := range words {
        wordCount[w]++
    }
    res := []int{}
    for i := 0; i < wordLen; i++ {
        left, count := i, 0
        seen := make(map[string]int)
        for j := i; j+wordLen <= len(s); j += wordLen {
            word := s[j:j+wordLen]
            if _, ok := wordCount[word]; ok {
                seen[word]++
                count++
                for seen[word] > wordCount[word] {
                    leftWord := s[left : left+wordLen]
                    seen[leftWord]--
                    left += wordLen
                    count--
                }
                if count == len(words) {
                    res = append(res, left)
                }
            } else {
                seen = make(map[string]int)
                count = 0
                left = j + wordLen
            }
        }
    }
    return res
}

File: sol/solution/0001-0100/0030. Substring with Concatenation of All Words/Solution.php
class Solution {
    /**
     * @param String $s
     * @param String[] $words
     * @return Integer[]
     */
    function findSubstring($s, $words) {
        if (!$s || count($words) == 0) return [];
        $wordLen = strlen($words[0]);
        $totalLen = $wordLen * count($words);
        $wordCount = array_count_values($words);
        $res = [];
        for ($i = 0; $i < $wordLen; $i++) {
            $left = $i; $count = 0; $seen = [];
            for ($j = $i; $j + $wordLen <= strlen($s); $j += $wordLen) {
                $word = substr($s, $j, $wordLen);
                if (isset($wordCount[$word])) {
                    $seen[$word] = ($seen[$word] ?? 0) + 1;
                    $count++;
                    while ($seen[$word] > $wordCount[$word]) {
                        $leftWord = substr($s, $left, $wordLen);
                        $seen[$leftWord]--;
                        $left += $wordLen;
                        $count--;
                    }
                    if ($count == count($words)) $res[] = $left;
                } else {
                    $seen = [];
                    $count = 0;
                    $left = $j + $wordLen;
                }
            }
        }
        return $res;
    }
}